File: apps/android/app/src/main/java/sh/zeron/android/data/UploadStash.kt
package sh.zeron.android.data

import sh.zeron.android.AppContextHolder
import java.io.File

/**
 * Durable staging for queued attachments (iOS UploadStash): the bytes of every
 * pending://-referenced attachment persist on device, keyed by uploadId, so
 * the delivery escort can re-derive its transfers after a retry, a relaunch,
 * or a tapped \"retry\" — the command in the doc names the ref; the stash holds
 * the bytes it points at.
 */
object UploadStash {
    const val PENDING_REF_PREFIX = "pending://"

    /** Test seam — JVM unit tests point this at a temp dir. */
    @Volatile
    var testDirectory: File? = null

    private fun directory(): File {
        val base = testDirectory ?: File(AppContextHolder.context?.filesDir ?: File("."), "zeron-uploads")
        base.mkdirs()
        return base
    }

    /** The persisted ref (uploads.rs PENDING_REF_PREFIX): the ORIGINAL file name rides it. */
    fun pendingRef(uploadId: String, name: String): String =
        "$PENDING_REF_PREFIX$uploadId/$name"

    /** Parse a pending ref back into (uploadId, fileName). */
    fun parseRef(ref: String): Pair<String, String>? {
        if (!ref.startsWith(PENDING_REF_PREFIX)) return null
        val body = ref.removePrefix(PENDING_REF_PREFIX)
        val slash = body.indexOf('/')
        if (slash <= 0 || slash == body.length - 1) return null
        val id = body.substring(0, slash)
        val name = body.substring(slash + 1)
        return if (id.isEmpty() || name.isEmpty()) null else id to name
    }

    private fun file(uploadId: String): File {
        val safe = uploadId.filter { it.isLetterOrDigit() || it == '-' }
        return File(directory(), "$safe.bin")
    }

    fun save(uploadId: String, data: ByteArray) {
        runCatching { file(uploadId).writeBytes(data) }
    }

    fun load(uploadId: String): ByteArray? = runCatching { file(uploadId).readBytes() }.getOrNull()

    fun delete(uploadId: String) {
        runCatching { file(uploadId).delete() }
    }

    /** Drop entries older than the command TTL (24h) — their commands expired. */
    fun sweep(ttlMs: Long = 24 * 60 * 60 * 1000L) {
        val cutoff = System.currentTimeMillis() - ttlMs
        directory().listFiles()?.forEach { f ->
            if (f.lastModified() < cutoff) f.delete()
        }
    }
}

File: apps/ios/Zeron/Sync/UploadStash.swift
// Durable staging for queued attachments (PR #168's "a send is a local
// write"): the bytes of every pending://-referenced attachment persist on
// device, keyed by uploadId, so the delivery escort can re-derive its
// transfers after a retry, a relaunch, or a tapped "retry" — the command in
// the doc names the ref; the stash holds the bytes it points at.
//
// Lives under the owning profile's DocDisk directory for sign-out cleanup.
// Entries are deleted when the host commits the upload; anything older than
// the command TTL (24h) is swept on demand.

import Foundation

enum UploadStash {
    static let pendingRefPrefix = "pending://"

    static func directory(namespace: String) -> URL {
        let base = DocDisk.directory(namespace: namespace).appendingPathComponent("uploads", isDirectory: true)
        try? FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
        return base
    }

    /// The persisted ref (uploads.rs PENDING_REF_PREFIX): the ORIGINAL file
    /// name rides the ref; both ends sanitize at resolution.
    static func pendingRef(uploadId: String, name: String) -> String {
        "\(pendingRefPrefix)\(uploadId)/\(name)"
    }

    /// Parse a pending ref back into (uploadId, fileName).
    static func parseRef(_ ref: String) -> (uploadId: String, name: String)? {
        guard ref.hasPrefix(pendingRefPrefix) else { return nil }
        let body = ref.dropFirst(pendingRefPrefix.count)
        guard let slash = body.firstIndex(of: "/") else { return nil }
        let id = String(body[..<slash])
        let name = String(body[body.index(after: slash)...])
        guard !id.isEmpty, !name.isEmpty else { return nil }
        return (id, name)
    }

    private static func url(uploadId: String, namespace: String) -> URL {
        let safe = uploadId.filter { $0.isLetter || $0.isNumber || $0 == "-" }
        return directory(namespace: namespace).appendingPathComponent("\(safe).bin")
    }

    static func save(uploadId: String, data: Data, namespace: String) {
        try? data.write(to: url(uploadId: uploadId, namespace: namespace), options: .atomic)
    }

    static func load(uploadId: String, namespace: String) -> Data? {
        try? Data(contentsOf: url(uploadId: uploadId, namespace: namespace))
    }

    static func delete(uploadId: String, namespace: String) {
        try? FileManager.default.removeItem(at: url(uploadId: uploadId, namespace: namespace))
    }

    /// Drop entries older than the command TTL — their commands have expired,
    /// so no escort will ever want the bytes again.
    static func sweep(ttlMs: Int64 = commandDefaultTtlMs, namespace: String) {
        let fm = FileManager.default
        guard let files = try? fm.contentsOfDirectory(
            at: directory(namespace: namespace), includingPropertiesForKeys: [.contentModificationDateKey]) else { return }
        let cutoff = Date(timeIntervalSinceNow: -Double(ttlMs) / 1000)
        for file in files {
            let modified = (try? file.resourceValues(forKeys: [.contentModificationDateKey])
                .contentModificationDate) ?? .distantPast
            if modified < cutoff {
                try? fm.removeItem(at: file)
            }
        }
    }
}
